compiler/src/vm/heap/object.rs

```diff
@@ -0,0 +1,257 @@
+                continue;
+            }
+            if existing_key.equals(heap, key) {
+                *existing_value = value;
+                return;
+            }
+        }
+        let entry = (hash, key, value);
+        if let Some(index) = self
+            .fields
+            .iter()
+            .position(|(the_hash, _, _)| *the_hash > hash)
+        {
+            self.fields.insert(index, entry);
+        } else {
+            self.fields.push(entry);
+        }
+    }
+    pub fn get(&self, heap: &Heap, key: Pointer) -> Option<Pointer> {
+        let hash = key.hash(heap);
+        Some(
+            self.fields
+                .iter()
+                .find(|(field_hash, field_key, _)| {
+                    hash == *field_hash && key.equals(heap, *field_key)
+                })?
+                .2,
+        )
+    }
+    fn len(&self) -> usize {
+        self.fields.len()
+    }
+    pub fn iter(&self) -> impl Iterator<Item = (Pointer, Pointer)> {
+        self.fields
+            .clone()
+            .into_iter()
+            .map(|(_, key, value)| (key, value))
+    }
+    fn equals(&self, heap: &Heap, other: &Struct) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        for (key_a, value_a) in self.iter() {
+            for (key_b, value_b) in other.iter() {
+                if key_a.equals(heap, key_b) && !value_a.equals(heap, value_b) {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+}
+
+impl Closure {
+    pub fn of_lir(module: Module, lir: Lir) -> Self {
+        Closure {
+            captured: vec![],
+            num_args: 0,
+            body: vec![
+                Instruction::TraceModuleStarts { module },
+                Instruction::CreateClosure {
+                    captured: vec![],
+                    num_args: 0,
+                    body: lir.instructions,
+                },
+                Instruction::Call { num_args: 0 },
+                Instruction::TraceModuleEnds,
+                Instruction::Return,
+            ],
+        }
+    }
+    pub fn of_module(db: &Database, module: Module) -> Option<Self> {
+        let lir = db.lir(module.clone())?;
+        Some(Self::of_lir(module, (*lir).clone()))
+    }
+}
+
+impl Data {
+    fn hash(&self, heap: &Heap) -> u64 {
+        let mut state = DefaultHasher::new();
+        self.hash_with_state(heap, &mut state);
+        state.finish()
+    }
+
+    fn hash_with_state<H: Hasher>(&self, heap: &Heap, state: &mut H) {
+        match self {
+            Data::Int(int) => int.value.hash(state),
+            Data::Text(text) => text.value.hash(state),
+            Data::Symbol(symbol) => symbol.value.hash(state),
+            Data::Struct(struct_) => {
+                let mut s = 0;
+                for (key, value) in struct_.iter() {
+                    s ^= key.hash(heap);
+                    s ^= value.hash(heap);
```

Comment:
Shouldn't this be something like `key.hash_with_state(heap, state)` instead of manually XORing the hashes?

Reply:
I tried to make sure that the ordering of the fields doesn't matter, because multiple structs considered equal may not return items in the same order (two keys with the same hash may be inserted in different orders)